File: tests/oracle/heritage_ownership_1204.cc
/*
 * HERITAGE-OWNERSHIP-0001: locked Ghidra 12.0.4 Funcdata::opHeritage /
 * Heritage::heritage ownership-boundary oracle.
 *
 * The fixture drives THREE consecutive `fd.opHeritage()` boundary calls
 * (pass 0 -> 1 -> 2 -> 3) on the same Funcdata through the production
 * wrapper (funcdata.hh:462 -> Heritage::heritage, heritage.cc:2663-2758)
 * and observes the ownership-boundary state after every pass: the pass
 * counter (public Funcdata::getHeritagePass, funcdata.hh:231) and, for the
 * inert SSA cases, the complete op-list / Varnode-bank projection through
 * the public PcodeOpBank/VarnodeBank iteration APIs.
 *
 * Observation-scope note (mirrored in the metadata envelope): the Ghidra
 * classes Funcdata and Heritage declare their state members in the
 * UNLABELED default-private access region, so a `#define private public`
 * fixture shim cannot expose `Heritage::maxdepth` / `globaldisjoint`.
 * Those fields are therefore Rust-asserted in the comparand fixture and
 * source-verified here (heritage.cc:218-224 ctor maxdepth=-1;
 * heritage.cc:2676-2772 rebuilds the ADT exactly when maxdepth==-1, i.e.
 * once at pass 1, and LocationMap merges are content-identical); they are
 * not part of the byte-compared stdout.
 *
 * Cases:
 *   empty_entry           one entry block, no IR; proves the pass counter
 *                         advances exactly once per boundary call and the
 *                         IR is untouched.
 *   register_free_promote a free register read SSA'd by canonical rename
 *                         (input promotion + def-use rewiring + free
 *                         deletion); full byte-compared state projection.
 *   phi_cycle_selfref     MULTIEQUAL self-reference cycle (cover-fixture
 *                         slot2 topology with a loop-carried def-use cycle
 *                         m_out -> r -> t3 -> m). Restricted projection:
 *                         this is the no-hang witness; Rugra's
 *                         placeMultiequals still places a redundant
 *                         parentless phi here (registered
 *                         HERITAGE-ADT-RENAME-0001), so the op/vn
 *                         projection is not byte-compared yet.
 */

#include <bits/stdc++.h>

#include "architecture.hh"
#include "cover.hh"
#include "database.hh"
#include "funcdata.hh"
#include "heritage.hh"
#include "libdecomp.hh"
#include "marshal.hh"
#include "op.hh"
#include "translate.hh"
#include "type.hh"
#include "typeop.hh"
#include "varnode.hh"

using namespace ghidra;

// Test-only observation shim. Funcdata's private section is an UNLABELED
// default-private region, so the `#define private public` trick used by
// other fixtures cannot expose `fd.heritage`. This shim re-declares the
// exact locked-oracle member sequence of Funcdata (funcdata.hh:57-95) in
// the same order with the same complete types, so a reinterpret_cast can
// reach the Heritage member to perform the one production startProcessing
// step a synthetic-graph fixture cannot reach otherwise:
// `heritage.buildInfoList()` (funcdata.cc:166). Layout is pinned to oracle
// commit e40ed13014025f82488b1f8f7bca566894ac376b, verified by the runner.
struct FuncdataHeritageShim {
  uint4 flags;
  uint4 clean_up_index;
  uint4 high_level_index;
  uint4 cast_phase_index;
  uint4 minLanedSize;
  int4 size;
  Architecture *glb;
  FunctionSymbol *functionSymbol;
  string name;
  string displayName;
  Address baseaddr;
  FuncProto funcp;
  ScopeLocal *localmap;
  vector<FuncCallSpecs *> qlst;
  vector<JumpTable *> jumpvec;
  VarnodeBank vbank;
  PcodeOpBank obank;
  BlockGraph bblocks;
  BlockGraph sblocks;
  Heritage heritage;
};

static Heritage &heritageOf(Funcdata &fd)
{
  return reinterpret_cast<FuncdataHeritageShim *>(&fd)->heritage;
}

class FixtureTranslate final : public Translate {
  VarnodeData dummy_register;

public:
  FixtureTranslate() {
    setBigEndian(false);
    setUniqueBase(0);
    insertSpace(new ConstantSpace(this, this));
    insertSpace(new AddrSpace(this, this, IPTR_PROCESSOR, "other", false, 8, 1, 1,
                              AddrSpace::hasphysical, 0, 0));
    insertSpace(new UniqueSpace(this, this, 2, 0));
    AddrSpace *ram = new AddrSpace(this, this, IPTR_PROCESSOR, "ram", false,
                                   8, 1, 3, AddrSpace::hasphysical, 0, 0);
    insertSpace(ram);
    AddrSpace *reg = new AddrSpace(this, this, IPTR_PROCESSOR, "register", false,
                                   8, 1, 4, AddrSpace::hasphysical, 0, 0);
    insertSpace(reg);
    SpacebaseSpace *stack = new SpacebaseSpace(
        this, this, "stack", 5, 8, ram, 1, true);
    insertSpace(stack);
    VarnodeData stack_pointer;
    stack_pointer.space = reg;
    stack_pointer.offset = 0;
    stack_pointer.size = 8;
    addSpacebasePointer(stack, stack_pointer, 8, true);
    insertSpace(new AddrSpace(this, this, IPTR_PROCESSOR, "join", false, 8, 1, 6,
                              AddrSpace::hasphysical, 0, 0));
    insertSpace(new IopSpace(this, this, 7));
    setDefaultCodeSpace(3);
    dummy_register.space = getSpace(4);
    dummy_register.offset = 0;
    dummy_register.size = 8;
  }

  void initialize(DocumentStorage &) override {}
  const VarnodeData &getRegister(const std::string &) const override {
    return dummy_register;
  }
  std::string getRegisterName(AddrSpace *, uintb, int4) const override { return ""; }
  std::string getExactRegisterName(AddrSpace *, uintb, int4) const override { return ""; }
  void getAllRegisters(std::map<VarnodeData, std::string> &) const override {}
  void getUserOpNames(std::vector<std::string> &) const override {}
  int4 instructionLength(const Address &) const override { return 0; }
  int4 oneInstruction(PcodeEmit &, const Address &) const override { return 0; }
  int4 printAssembly(AssemblyEmit &, const Address &) const override { return 0; }
};

class FixtureArchitecture final : public Architecture {
protected:
  Translate *buildTranslator(DocumentStorage &) override { return (Translate *)0; }
  void buildLoader(DocumentStorage &) override {}
  PcodeInjectLibrary *buildPcodeInjectLibrary(void) override {
    return (PcodeInjectLibrary *)0;
  }
  void buildTypegrp(DocumentStorage &) override {}
  void buildCoreTypes(DocumentStorage &) override {}
  void buildCommentDB(DocumentStorage &) override {}
  void buildStringManager(DocumentStorage &) override {}
  void buildConstantPool(DocumentStorage &) override {}
  void buildContext(DocumentStorage &) override {}
  void buildSymbols(DocumentStorage &) override {}
  void buildSpecFile(DocumentStorage &) override {}
  void modifySpaces(Translate *) override {}
  void resolveArchitecture(void) override {}

public:
  FixtureArchitecture() {
    FixtureTranslate *fixture_translate = new FixtureTranslate();
    translate = fixture_translate;
    copySpaces(fixture_translate);
    max_basetype_size = 16;
    types = new TypeFactory(this);
    types->setupSizes();
    types->setCoreType("xunknown1", 1, TYPE_UNKNOWN, false);
    types->setCoreType("xunknown2", 2, TYPE_UNKNOWN, false);
    types->setCoreType("xunknown4", 4, TYPE_UNKNOWN, false);
    types->setCoreType("xunknown8", 8, TYPE_UNKNOWN, false);
    types->cacheCoreTypes();
    TypeOp::registerInstructions(inst, types, translate);
    symboltab = new Database(this, false);
    symboltab->attachScope(new ScopeInternal(0x101, "", this), (Scope *)0);
    ProtoModel *model = new ProtoModel(this);
    std::istringstream stream(
        "<prototype name=\"fixture\" extrapop=\"0\"><input/><output/></prototype>");
    XmlDecode decoder(this);
    decoder.ingestStream(stream);
    model->decode(decoder);
    protoModels[model->getName()] = model;
    setDefaultModel(model);
  }

  void printMessage(const std::string &) const override {}
};

// Opcode abbreviation shared with the Rust comparand (both sides map their
// own enum; only the fixture's five opcodes are needed).
static const char *opcodeName(OpCode opc)
{
  switch (opc) {
  case CPUI_COPY: return "COPY";
  case CPUI_INT_ADD: return "INT_ADD";
  case CPUI_INT_OR: return "INT_OR";
  case CPUI_INT_MULT: return "INT_MULT";
  case CPUI_MULTIEQUAL: return "MULTIEQUAL";
  default: return "OTHER";
  }
}

// Varnode descriptor shared with the Rust comparand:
//   constant  -> C<size>
//   register  -> R<hex-offset>:<I|W|F>
//   unique    -> U<size>:<I|W|F>
//   stack     -> S<hex-offset>:<I|W|F>
//   ram       -> M<hex-offset>:<I|W|F>
// Absolute unique offsets are deliberately not printed (allocation bases
// differ between the two implementations); the identity-relevant state
// (size, def-use class) is.
static std::string vnDescriptor(const Varnode *vn)
{
  std::ostringstream out;
  AddrSpace *spc = vn->getSpace();
  if (vn->isConstant()) {
    out << 'C' << vn->getSize();
    return out.str();
  }
  char code = 'X';
  if (spc->getType() == IPTR_SPACEBASE)
    code = 'S';
  else if (spc->getName() == "register")
    code = 'R';
  else if (spc->getType() == IPTR_INTERNAL)
    code = 'U';
  else if (spc->getName() == "ram")
    code = 'M';
  out << code;
  if (code == 'R' || code == 'S' || code == 'M')
    out << hex << vn->getOffset() << dec;
  else
    out << vn->getSize();
  out << ':';
  if (vn->isInput())
    out << 'I';
  else if (vn->isWritten())
    out << 'W';
  else
    out << 'F';
  return out.str();
}

class Graph {
public:
  Funcdata fd;
  FixtureArchitecture &arch;
  AddrSpace *ram;
  AddrSpace *reg;
  std::map<PcodeOp *, std::string> opNames;
  int4 nextOffset;

  Graph(FixtureArchitecture &a, const std::string &name, uintb base)
    : fd(name, name, a.symboltab->getGlobalScope(), Address(a.getSpace(3), base),
         (FunctionSymbol *)0, 0x20),
      arch(a), ram(a.getSpace(3)), reg(a.getSpace(4)), nextOffset(0) {}

  BlockBasic *makeBlock(void)
  {
    BlockGraph &blocks = const_cast<BlockGraph &>(fd.getBasicBlocks());
    BlockBasic *block = blocks.newBlockBasic(&fd);
    // Production assigns FlowBlock::index via BlockGraph::findSpanningTree
    // reverse post-order (block.cc:1081), which this fixture does not run.
    // The creation-order indices here are a deterministic, distinct
    // precondition matching the Rust comparand's explicit index argument.
    return block;
  }

  void edge(BlockBasic *from, BlockBasic *to)
  {
    BlockGraph &blocks = const_cast<BlockGraph &>(fd.getBasicBlocks());
    blocks.addEdge(from, to);
  }

  PcodeOp *makeOp(const std::string &name, OpCode opcode, int4 inputs)
  {
    Address pc(ram, fd.getAddress().getOffset() + (uintb)nextOffset);
    nextOffset += 1;
    PcodeOp *op = fd.newOp(inputs, pc);
    fd.opSetOpcode(op, opcode);
    opNames.insert(std::make_pair(op, name));
    return op;
  }

  Varnode *uniqueOut(int4 size, PcodeOp *op)
  {
    return fd.newUniqueOut(size, op);
  }

  Varnode *constant(int4 size, uintb value)
  {
    return fd.newConstant(size, value);
  }

  Varnode *freeRegister(uintb offset, int4 size)
  {
    return fd.newVarnode(size, Address(reg, offset));
  }

  Varnode *writtenRegister(uintb offset, int4 size, PcodeOp *op)
  {
    Varnode *vn = fd.newVarnode(size, Address(reg, offset));
    fd.opSetOutput(op, vn);
    return op->getOut();
  }

  void setInput(PcodeOp *op, Varnode *vn, int4 slot)
  {
    fd.opSetInput(op, vn, slot);
  }

  void insertEnd(PcodeOp *op, BlockBasic *block)
  {
    fd.opInsertEnd(op, block);
  }

  // Production pre-state: Funcdata::structureReset (funcdata_block.cc:703)
  // computes loop structure + forward dominators before ActionHeritage;
  // those two BlockGraph steps are public and are applied directly here.
  void prepareStructure(void)
  {
    vector<FlowBlock *> rootlist;
    BlockGraph &blocks = const_cast<BlockGraph &>(fd.getBasicBlocks());
    blocks.structureLoops(rootlist);
    blocks.calcForwardDominator(rootlist);
    // startProcessing's heritage step (funcdata.cc:166): the per-space
    // HeritageInfo list is built BEFORE the first ActionHeritage pass.
    heritageOf(fd).buildInfoList();
  }

  // Three consecutive opHeritage boundary calls, recording the public pass
  // counter after each pass.
  void runThreePasses(int4 *passSeq)
  {
    prepareStructure();
    for (int4 i = 0; i < 3; ++i) {
      fd.opHeritage();
      passSeq[i] = fd.getHeritagePass();
    }
  }

  std::string opList(void)
  {
    std::ostringstream out;
    bool first = true;
    for (PcodeOpTree::const_iterator iter = fd.beginOpAll(); iter != fd.endOpAll(); ++iter) {
      const PcodeOp *op = (*iter).second;
      if (!first)
        out << ';';
      first = false;
      std::map<PcodeOp *, std::string>::const_iterator name =
          opNames.find(const_cast<PcodeOp *>(op));
      out << (name == opNames.end() ? std::string("phi") : name->second)
          << '.' << opcodeName(op->code()) << '(';
      for (int4 slot = 0; slot < op->numInput(); ++slot) {
        if (slot != 0)
          out << ',';
        out << vnDescriptor(op->getIn(slot));
      }
      out << ')';
    }
    return out.str();
  }

  int4 opCount(void)
  {
    int4 count = 0;
    for (PcodeOpTree::const_iterator iter = fd.beginOpAll(); iter != fd.endOpAll(); ++iter)
      count += 1;
    return count;
  }

  std::string vnList(void)
  {
    std::vector<std::string> parts;
    for (VarnodeLocSet::const_iterator iter = fd.beginLoc(); iter != fd.endLoc(); ++iter)
      parts.push_back(vnDescriptor(*iter));
    std::sort(parts.begin(), parts.end());
    std::ostringstream out;
    for (size_t i = 0; i < parts.size(); ++i) {
      if (i != 0)
        out << ',';
      out << parts[i];
    }
    return out.str();
  }

  int4 vnCount(void)
  {
    int4 count = 0;
    for (VarnodeLocSet::const_iterator iter = fd.beginLoc(); iter != fd.endLoc(); ++iter)
      count += 1;
    return count;
  }
};

static void runEmptyEntry(FixtureArchitecture &arch)
{
  Graph g(arch, "empty_entry", 0x5000);
  g.makeBlock();
  int4 passSeq[3];
  g.runThreePasses(passSeq);
  std::cout << "case=empty_entry"
            << "|pass=" << passSeq[0] << ',' << passSeq[1] << ',' << passSeq[2]
            << "|ops=" << g.opCount()
            << "|vns=" << g.vnCount() << '\n';
}

static void runRegisterFreePromote(FixtureArchitecture &arch)
{
  Graph g(arch, "register_free_promote", 0x5100);
  BlockBasic *b0 = g.makeBlock();
  BlockBasic *b1 = g.makeBlock();
  g.edge(b0, b1);
  // Two separate size-4 constants: Funcdata::opSetInput clones a constant
  // that already has a descendant (funcdata_op.cc opSetInput), so the two
  // readers get distinct constants exactly as raw p-code would.
  Varnode *c4a = g.constant(4, 7);
  Varnode *c4b = g.constant(4, 7);
  Varnode *c8 = g.constant(8, 5);
  // Raw p-code creates a distinct free Varnode per read of the same
  // register (Ghidra invariant: a free Varnode has at most one descendant,
  // varnode.cc addDescend).
  Varnode *f1 = g.freeRegister(0x30, 8);
  Varnode *f2 = g.freeRegister(0x30, 8);
  PcodeOp *d1 = g.makeOp("d1", CPUI_INT_ADD, 2);
  g.setInput(d1, f1, 0);
  g.setInput(d1, c4a, 1);
  g.writtenRegister(0x30, 8, d1);
  g.insertEnd(d1, b0);
  PcodeOp *w2 = g.makeOp("w2", CPUI_COPY, 1);
  g.setInput(w2, c8, 0);
  g.uniqueOut(8, w2);
  g.insertEnd(w2, b0);
  PcodeOp *r1 = g.makeOp("r1", CPUI_INT_OR, 2);
  g.setInput(r1, f2, 0);
  g.setInput(r1, c4b, 1);
  g.uniqueOut(8, r1);
  g.insertEnd(r1, b1);
  int4 passSeq[3];
  g.runThreePasses(passSeq);
  std::cout << "case=register_free_promote"
            << "|pass=" << passSeq[0] << ',' << passSeq[1] << ',' << passSeq[2]
            << "|ops=" << g.opCount()
            << "|oplist=" << g.opList()
            << "|vns=" << g.vnCount()
            << "|vnlist=" << g.vnList() << '\n';
}

static void runPhiCycleSelfRef(FixtureArchitecture &arch)
{
  Graph g(arch, "phi_cycle_selfref", 0x5200);
  BlockBasic *b0 = g.makeBlock();
  BlockBasic *b1 = g.makeBlock();
  BlockBasic *b2 = g.makeBlock();
  BlockBasic *b3 = g.makeBlock();
  g.edge(b0, b1);
  g.edge(b1, b2);
  g.edge(b2, b1);
  g.edge(b2, b3);
  Varnode *c8 = g.constant(8, 5);
  Varnode *c4 = g.constant(4, 7);
  PcodeOp *d0 = g.makeOp("d0", CPUI_COPY, 1);
  g.setInput(d0, c8, 0);
  g.uniqueOut(8, d0);
  g.insertEnd(d0, b0);
  Varnode *f = g.freeRegister(0x28, 8);
  PcodeOp *m = g.makeOp("m", CPUI_MULTIEQUAL, 2);
  Varnode *mOut = g.writtenRegister(0x28, 8, m);
  g.setInput(m, f, 0);
  PcodeOp *r = g.makeOp("r", CPUI_INT_ADD, 2);
  Varnode *t3 = g.writtenRegister(0x28, 8, r);
  g.setInput(r, mOut, 0);
  g.setInput(r, c4, 1);
  g.setInput(m, t3, 1);
  g.insertEnd(m, b1);
  g.insertEnd(r, b2);
  PcodeOp *o = g.makeOp("o", CPUI_INT_OR, 2);
  g.setInput(o, mOut, 0);
  g.setInput(o, c4, 1);
  g.uniqueOut(8, o);
  g.insertEnd(o, b3);
  int4 passSeq[3];
  g.runThreePasses(passSeq);
  // Restricted projection: the no-hang witness. Reaching this print proves
  // all three boundary calls returned on the phi self-reference cycle.
  std::cout << "case=phi_cycle_selfref"
            << "|pass=" << passSeq[0] << ',' << passSeq[1] << ',' << passSeq[2]
            << "|completed=1" << '\n';
}

int main(void)
{
  std::cout << std::unitbuf;
  std::vector<std::string> spec_paths;
  startDecompilerLibrary(spec_paths);
  std::cout << "schema=1|fixture=HERITAGE-OWNERSHIP-0001|oracle=e40ed13014025f82488b1f8f7bca566894ac376b\n";
  try {
    FixtureArchitecture arch;
    runEmptyEntry(arch);
    runRegisterFreePromote(arch);
    runPhiCycleSelfRef(arch);
  }
  catch (const std::exception &error) {
    std::cerr << "fixture error: " << error.what() << '\n';
    return 1;
  }
  return 0;
}

File: tests/oracle/heritage_ownership_1204.rs
// HERITAGE-OWNERSHIP-0001: Rugra comparand for the locked Ghidra 12.0.4
// Funcdata::opHeritage / Heritage::heritage ownership-boundary oracle.
// Mirrors tests/oracle/heritage_ownership_1204.cc case for case: the same
// synthetic CFG/def-use graphs are built through the production Funcdata
// APIs, three consecutive `op_heritage` boundary calls (pass 0 -> 1 -> 2
// -> 3) run on the same `&mut Funcdata`, and the complete state
// projection (pass counter, op list with per-slot input classes,
// Varnode-bank multiset) is printed in the shared observation format.
//
// The Rust fixture additionally ASSERTS (before printing) the
// ownership-boundary internals the C++ side cannot print because Ghidra
// declares them in an unlabeled default-private region:
//   - maxdepth == -1 at construction and exactly one ADT rebuild across
//     the three passes (heritage.cc:218-224 / 2676-2677),
//   - the persistent globaldisjoint cover survives every mem::take
//     round-trip of the Heritage object.

use std::sync::Arc;

use rugra::address::Address;
use rugra::block::{BlockBasic, FlowBlock};
use rugra::funcdata::Funcdata;
use rugra::op::PcodeOp;
use rugra::opcodes::OpCode;
use rugra::space::AddressSpace;
use rugra::varnode::Varnode;

type BlockRef = Arc<std::sync::RwLock<dyn FlowBlock + Send + Sync>>;
type VnRef = Arc<std::sync::RwLock<Varnode>>;
type OpRef = Arc<std::sync::RwLock<PcodeOp>>;

fn opcode_name(opc: OpCode) -> &'static str {
    match opc {
        OpCode::CPUI_COPY => "COPY",
        OpCode::CPUI_INT_ADD => "INT_ADD",
        OpCode::CPUI_INT_OR => "INT_OR",
        OpCode::CPUI_INT_MULT => "INT_MULT",
        OpCode::CPUI_MULTIEQUAL => "MULTIEQUAL",
        _ => "OTHER",
    }
}

// Varnode descriptor shared with the C++ oracle fixture:
//   constant -> C<size>; register -> R<hex-offset>:<I|W|F>;
//   unique -> U<size>:<I|W|F>; stack -> S<off>:<..>; ram -> M<off>:<..>
fn vn_descriptor(vn: &VnRef) -> String {
    let value = vn.read().unwrap();
    if value.is_constant() {
        return format!("C{}", value.size);
    }
    let code = match value.address_space {
        AddressSpace::Stack => 'S',
        AddressSpace::Register => 'R',
        AddressSpace::Unique => 'U',
        AddressSpace::Ram => 'M',
        _ => 'X',
    };
    let head = if matches!(
        value.address_space,
        AddressSpace::Stack | AddressSpace::Register | AddressSpace::Ram
    ) {
        format!("{code}{}", format!("{:x}", value.loc.as_u64()))
    } else {
        format!("{code}{}", value.size)
    };
    let state = if value.is_input() {
        'I'
    } else if value.is_written() {
        'W'
    } else {
        'F'
    };
    format!("{head}:{state}")
}

struct Graph {
    fd: Funcdata,
    base: u64,
    next_offset: u64,
    ops: Vec<(OpRef, &'static str)>,
}

impl Graph {
    fn new(name: &str, base: u64) -> Self {
        Graph {
            fd: Funcdata::new(name, Address::new(base), 0x20),
            base,
            next_offset: 0,
            ops: Vec::new(),
        }
    }

    fn make_block(&mut self, index: i32) -> BlockRef {
        let block: BlockRef = Arc::new(std::sync::RwLock::new(BlockBasic::new(
            index,
            Address::new(self.base),
        )));
        self.fd.bblocks.add_block(block.clone());
        block
    }

    fn edge(&mut self, from: &BlockRef, to: &BlockRef) {
        self.fd.bblocks.add_edge(from.clone(), to.clone());
    }

    fn make_op(&mut self, name: &'static str, opcode: OpCode, inputs: usize) -> OpRef {
        let pc = Address::new(self.base + self.next_offset);
        self.next_offset += 1;
        let op = self.fd.new_op(inputs, pc);
        self.fd.op_set_opcode(&op, opcode);
        self.ops.push((op.0.clone(), name));
        op.0.clone()
    }

    fn unique_out(&mut self, size: usize, op: &OpRef) -> VnRef {
        self.fd
            .new_unique_out(size, &rugra::op::PcodeOpRef(op.clone()))
    }

    fn constant(&mut self, size: usize, value: u64) -> VnRef {
        self.fd.new_constant(size, value)
    }

    fn free_register(&mut self, offset: u64, size: usize) -> VnRef {
        self.fd
            .vbank
            .create_with_space(size, AddressSpace::Register, offset)
    }

    fn written_register(&mut self, offset: u64, size: usize, op: &OpRef) -> VnRef {
        let vn = self
            .fd
            .vbank
            .create_with_space(size, AddressSpace::Register, offset);
        self.fd.op_set_output(&rugra::op::PcodeOpRef(op.clone()), vn.clone());
        vn
    }

    fn set_input(&mut self, op: &OpRef, vn: &VnRef, slot: usize) {
        self.fd
            .op_set_input(&rugra::op::PcodeOpRef(op.clone()), vn.clone(), slot);
    }

    fn insert_end(&mut self, op: &OpRef, block: &BlockRef) {
        self.fd
            .op_insert_end(&rugra::op::PcodeOpRef(op.clone()), block);
    }

    fn label_of(&self, op: &OpRef) -> &'static str {
        for (candidate, name) in &self.ops {
            if Arc::ptr_eq(candidate, op) {
                return name;
            }
        }
        "phi"
    }

    /// Production pre-state: `Funcdata::startProcessing`
    /// (funcdata.cc:150-167) runs structureReset (loop structure + forward
    /// dominators) and builds the Heritage info list (funcdata.cc:166)
    /// before the first ActionHeritage pass. The synthetic-graph fixture
    /// applies the same two steps: Rugra's dominator producer is
    /// `BlockGraph::build_dom_tree`, and `build_info_list` is the direct
    /// buildInfoList port.
    fn prepare_structure(&mut self) {
        self.fd.bblocks.build_dom_tree();
        self.fd.heritage.build_info_list();
    }

    /// Three consecutive op_heritage boundary calls on the same `&mut
    /// Funcdata`, recording the public pass counter after each pass and
    /// asserting the Rust-side ownership internals.
    fn run_three_passes(&mut self, expect_maxdepth: i32, expect_gd: usize) -> Vec<i32> {
        self.prepare_structure();
        assert_eq!(self.fd.heritage.maxdepth, -1, "ctor sentinel maxdepth=-1");
        let mut seq = Vec::new();
        for _ in 0..3 {
            self.fd.op_heritage();
            seq.push(self.fd.num_heritage_passes());
            // The ADT rebuild sentinel fires on pass 1 only; maxdepth is
            // then constant across the remaining passes.
            assert_eq!(self.fd.heritage.maxdepth, expect_maxdepth);
            assert_eq!(
                self.fd.heritage.globaldisjoint.themap.len(),
                expect_gd,
                "persistent globaldisjoint must survive every mem::take round-trip"
            );
        }
        seq
    }

    fn op_list(&self) -> String {
        let mut parts = Vec::new();
        for op_ref in &self.fd.obank.optree {
            let op = op_ref.0.read().unwrap();
            let mut part = format!(
                "{}.{}(",
                self.label_of(&op_ref.0),
                opcode_name(op.opcode)
            );
            for (slot, input) in op.inrefs.iter().enumerate() {
                if slot != 0 {
                    part.push(',');
                }
                part.push_str(&vn_descriptor(input));
            }
            part.push(')');
            parts.push(part);
        }
        parts.join(";")
    }

    fn op_count(&self) -> usize {
        self.fd.obank.optree.len()
    }

    fn vn_list(&self) -> String {
        let mut parts: Vec<String> = self
            .fd
            .vbank
            .loc_tree
            .iter()
            .map(|entry| vn_descriptor(&entry.0))
            .collect();
        parts.sort();
        parts.join(",")
    }

    fn vn_count(&self) -> usize {
        self.fd.vbank.loc_tree.len()
    }
}

fn pass_text(seq: &[i32]) -> String {
    format!("{},{},{}", seq[0], seq[1], seq[2])
}

fn main() {
    println!("schema=1|fixture=HERITAGE-OWNERSHIP-0001|oracle=e40ed13014025f82488b1f8f7bca566894ac376b");

    // case=empty_entry: one entry block, no IR.
    {
        let mut g = Graph::new("empty_entry", 0x5000);
        g.make_block(0);
        let seq = g.run_three_passes(1, 0);
        println!(
            "case=empty_entry|pass={}|ops={}|vns={}",
            pass_text(&seq),
            g.op_count(),
            g.vn_count()
        );
    }

    // case=register_free_promote: canonical rename promotes the free reads
    // to one input, rewires the second reader onto the write, and deletes
    // the consumed frees.
    {
        let mut g = Graph::new("register_free_promote", 0x5100);
        let b0 = g.make_block(0);
        let b1 = g.make_block(1);
        g.edge(&b0, &b1);
        let c4a = g.constant(4, 7);
        let c4b = g.constant(4, 7);
        let c8 = g.constant(8, 5);
        let f1 = g.free_register(0x30, 8);
        let f2 = g.free_register(0x30, 8);
        let d1 = g.make_op("d1", OpCode::CPUI_INT_ADD, 2);
        g.set_input(&d1, &f1, 0);
        g.set_input(&d1, &c4a, 1);
        g.written_register(0x30, 8, &d1);
        g.insert_end(&d1, &b0);
        let w2 = g.make_op("w2", OpCode::CPUI_COPY, 1);
        g.set_input(&w2, &c8, 0);
        g.unique_out(8, &w2);
        g.insert_end(&w2, &b0);
        let r1 = g.make_op("r1", OpCode::CPUI_INT_OR, 2);
        g.set_input(&r1, &f2, 0);
        g.set_input(&r1, &c4b, 1);
        g.unique_out(8, &r1);
        g.insert_end(&r1, &b1);
        // Two-block chain: dominator depths 1,2 (block.cc:2056 root=1) ->
        // maxdepth 2; three disjoint covers (register 0x30, t2, t3).
        let seq = g.run_three_passes(2, 3);
        println!(
            "case=register_free_promote|pass={}|ops={}|oplist={}|vns={}|vnlist={}",
            pass_text(&seq),
            g.op_count(),
            g.op_list(),
            g.vn_count(),
            g.vn_list()
        );
    }

    // case=phi_cycle_selfref: MULTIEQUAL self-reference cycle (cover
    // fixture slot2 topology with the loop-carried def-use cycle
    // m_out -> r -> t3 -> m). Restricted projection: the no-hang witness.
    {
        let mut g = Graph::new("phi_cycle_selfref", 0x5200);
        let b0 = g.make_block(0);
        let b1 = g.make_block(1);
        let b2 = g.make_block(2);
        let b3 = g.make_block(3);
        g.edge(&b0, &b1);
        g.edge(&b1, &b2);
        g.edge(&b2, &b1);
        g.edge(&b2, &b3);
        let c8 = g.constant(8, 5);
        let c4 = g.constant(4, 7);
        let d0 = g.make_op("d0", OpCode::CPUI_COPY, 1);
        g.set_input(&d0, &c8, 0);
        g.unique_out(8, &d0);
        g.insert_end(&d0, &b0);
        let f = g.free_register(0x28, 8);
        let m = g.make_op("m", OpCode::CPUI_MULTIEQUAL, 2);
        let m_out = g.written_register(0x28, 8, &m);
        g.set_input(&m, &f, 0);
        let r = g.make_op("r", OpCode::CPUI_INT_ADD, 2);
        let t3 = g.written_register(0x28, 8, &r);
        g.set_input(&r, &m_out, 0);
        g.set_input(&r, &c4, 1);
        g.set_input(&m, &t3, 1);
        g.insert_end(&m, &b1);
        g.insert_end(&r, &b2);
        let o = g.make_op("o", OpCode::CPUI_INT_OR, 2);
        g.set_input(&o, &m_out, 0);
        g.set_input(&o, &c4, 1);
        g.unique_out(8, &o);
        g.insert_end(&o, &b3);
        // Four-block dominator chain b0 > b1 > b2 > b3: depths 1..4; the
        // persistent cover holds the merged register 0x28 entry plus the
        // two unique ranges.
        let seq = g.run_three_passes(4, 3);
        println!(
            "case=phi_cycle_selfref|pass={}|completed=1",
            pass_text(&seq)
        );
    }
}
